### PR TITLE
Certificate Signing Request

### PR DESCRIPTION
**Pull request by Lukasa**
This patch adds support for working with Certificate Signing Requests to
swift-certificates.

CSRs are a common feature of X.509 based PKIs. They're used to encapsulate
a number of details about the certificate being requested, and are
typically handed to CAs in exchange for a certificate. Their purpose is
to prove possession of a private key, and optionally some other useful
pieces of data.

This patch provides a complete representation of the CSR type, as well as
semantic types to cover the attributes that can be applied to a CSR.
Additionally, CSR signature validation is also available.

### CODE REVIEW

**Review by dnadoba — CHANGES_REQUESTED** [review]
Looks mostly good, see inline comments for a couple things I think we should change before merging.

File: Sources/X509/CSR/CSRAttributes.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    ///
+    /// - Parameter oid: The OID to search for.
+    @inlinable
+    public subscript(oid oid: ASN1ObjectIdentifier) -> CertificateSigningRequest.Attribute? {
+        get {
+            return self.first(where: { $0.oid == oid })
+        }
+        set {
+            if let newValue = newValue {
+                precondition(oid == newValue.oid)
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/X509/CSR/CSRAttributes.swift:131]: That feels fairly un-handy, and prevents us avoiding CoW in some cases.

File: Sources/X509/CSR/CSRAttribute.swift [hunk 2]
```diff
@@ -47,7 +47,34 @@ extension CertificateSigningRequest {
...
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self.oid)
+
+        // This achieves order-independent hashing without
+        // having to sort anything.
+        var hash = 0
+        for element in self.values {
+            var newHasher = Hasher()
+            element.hash(into: &newHasher)
+            hash ^= newHasher.finalize()
```
Comment (dnadoba) [inline Sources/X509/CSR/CSRAttribute.swift:72]: Nit: the XOR trick is neat but can we add a comment which explains why this does achieve order-independent hashing? E.g. something that says that any associative operation would do it?

**Review by dnadoba — COMMENTED** [review]
Looks great minus one nit

File: Sources/X509/CSR/CSRAttribute.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+}
+
+extension CertificateSigningRequest.Attribute: Hashable {
+    @inlinable
+    public static func ==(lhs: CertificateSigningRequest.Attribute, rhs: CertificateSigningRequest.Attribute) -> Bool {
+        if lhs.oid != rhs.oid { return false }
+
+        for element in lhs.values {
```
Comment (dnadoba) [inline Sources/X509/CSR/CSRAttribute.swift:55]: `lhs.values.count == rhs.values.count` must be true as well or they will be equal even if `rhs` has additional elements.
Can we add for it as well?

Same needs to be done for `CertificateSigningRequest.Attributes`

**Review by dnadoba — CHANGES_REQUESTED** [review]
Catched something on a second look

File: Sources/X509/CSR/CSRAttribute.swift (hunk 3 above)
Comment (Lukasa) [reply Sources/X509/CSR/CSRAttribute.swift:55]: Oh good catch!

**Comment by Lukasa** [comment]
Addressed those comments, thanks!

**Review by dnadoba — APPROVED** [review]
LGTM

**Pull request merged**